### ISSUE TITLE
Could not create SSL/TLS secure channel Powershell 5?

### ISSUE THREAD

**Issue by orionneo** [pspete/psPAS#110]
## Your Environment

* PowerShell Version: 5.0
* psPAS Branch & Version: most recent one.
* CyberArk Version: 10.4

## Expected Behaviour

Authentication via New-PASSession should work 100%.

## Current Behaviour

When we are attempting to authenticate via the following command:

$token = New-PASSession -Credential $Credentials -BaseURI $URL

The following error is reported:The request was aborted: Could not create SSL/TLS secure channel.
Invoke-PASRestMethod : The request was aborted: Could not create SSL/TLS secure channel.
At C:\Users\za78ca\Documents\WindowsPowerShell\Modules\psPAS\Functions\Authentication\New-PASSharedSession.ps1
:91 char:18
+ ... ASSession = Invoke-PASRestMethod -Uri $URI -Method POST -Body $Body - ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-PASRestMethod

It is not about certificate which is valid and verified even if we check the PVWA in our browsers.

The powershell version is the following: 
Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      15063  1387 

If I run the same command on PowerShell 6, just for testing, it works (on the same machine):

PS C:\Users\za78ca> $token

sessionToken    BaseURI
------------    -------
{Authorization} https://xxxxx

To summarize, the authentication works with powershell 6 but not with powershell 5. We would like to stick with PS 5, but we couldn't determine the root cause for this.... can you please support?

**Comment by pspete** [comment]
You have logged the issue against `New-PASSession`, but the error message you provide is related to `New-PASSharedSession` - can you clarify with which function you see the error?

**Comment by orionneo** [comment]
Hi Pete, the issue is happening to both!

**Comment by pspete** [comment]
you've set `bindingConfiguration="httpsBinding"` against the relevant authentication services in the web.config file?

**Comment by orionneo** [comment]
We have only Rewrite configuration and a rule to route http to https. What is weird is that with PS 6.0 it works! but not with PS 5.x~ Any hints?

**Comment by orionneo** [comment]
This is how it looks like:
 <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Shared.AuthenticationService">
        <endpoint address="" binding="basicHttpBinding" bindingConfiguration="defaultBinding" contract="CyberArk.WSAuthentication.Shared.IAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Cyberark.CyberArkAuthenticationService">
        <endpoint name="AuthEndpoint" address="/" behaviorConfiguration="web" binding="webHttpBinding" bindingConfiguration="httpsBinding" contract="CyberArk.WSAuthentication.Cyberark.ICyberArkAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.PasswordVault.WebServices.WF.PIMServices">
        <endpoint name="PIMEndpoint" address="/" behaviorConfiguration="web" binding="webHttpBinding" bindingConfiguration="httpsBinding" contract="CyberArk.PasswordVault.WebServices.WF.IPIMServices" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Shared.RestfulAuthenticationService">
        <endpoint address="" binding="webHttpBinding" bindingConfiguration="httpsBinding" behaviorConfiguration="web" contract="CyberArk.WSAuthentication.Shared.IRestfulAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.SAML.SAMLAuthenticationService">
        <endpoint address="" binding="webHttpBinding" bindingConfiguration="httpsBinding" behaviorConfiguration="web" contract="CyberArk.WSAuthentication.SAML.ISAMLAuthenticationService" />
      </service>
    </services>

**Comment by orionneo** [comment]
```
 <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Shared.AuthenticationService">
        <endpoint address="" binding="basicHttpBinding" bindingConfiguration="defaultBinding" contract="CyberArk.WSAuthentication.Shared.IAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Cyberark.CyberArkAuthenticationService">
        <endpoint name="AuthEndpoint" address="/" behaviorConfiguration="web" binding="webHttpBinding" bindingConfiguration="httpsBinding" contract="CyberArk.WSAuthentication.Cyberark.ICyberArkAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.PasswordVault.WebServices.WF.PIMServices">
        <endpoint name="PIMEndpoint" address="/" behaviorConfiguration="web" binding="webHttpBinding" bindingConfiguration="httpsBinding" contract="CyberArk.PasswordVault.WebServices.WF.IPIMServices" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.Shared.RestfulAuthenticationService">
        <endpoint address="" binding="webHttpBinding" bindingConfiguration="httpsBinding" behaviorConfiguration="web" contract="CyberArk.WSAuthentication.Shared.IRestfulAuthenticationService" />
      </service>
      <service behaviorConfiguration="defaultBehavior" name="CyberArk.WSAuthentication.SAML.SAMLAuthenticationService">
        <endpoint address="" binding="webHttpBinding" bindingConfiguration="httpsBinding" behaviorConfiguration="web" contract="CyberArk.WSAuthentication.SAML.ISAMLAuthenticationService" />
      </service>
    </services>
```

**Comment by pspete** [comment]
The module attempts to force use of TLS1.2 if it is available.
`[Net.SecurityProtocolType].GetEnumNames()` will give you the available security Protocols
`[Net.ServicePointManager]::SecurityProtocol` will show you what is currently in use.

Check the `[Net.ServicePointManager]::SecurityProtocol` output before and after attempting to authenticate in both PowerShell 5 & PowerShell Core to see if there is any difference in the protocol being used:

````powershell
PS>[Net.SecurityProtocolType].GetEnumNames()
SystemDefault
Ssl3
Tls
Tls11
Tls12

PS>[System.Net.ServicePointManager]::SecurityProtocol
Tls12
````

**Comment by orionneo** [comment]
Here are the results:

**Powershell Core:**

```
PS C:\Users\xxxxxa> [Net.SecurityProtocolType].GetEnumNames()
SystemDefault
Ssl3
Tls
Tls11
Tls12
PS C:\Users\xxxxa> [Net.ServicePointManager]::SecurityProtocol
SystemDefault

```
**Powershell 5**

```
PS C:\Users\xxxxxa> [Net.SecurityProtocolType].GetEnumNames()
SystemDefault
Ssl3
Tls
Tls11
Tls12
PS C:\Users\xxxxxa> [Net.ServicePointManager]::SecurityProtocol
Tls, Tls11, Tls12
```

**Comment by pspete** [comment]
In PSCore, does the following return a 200 response?

`Invoke-WebRequest -Uri https://YourPVWA/PasswordVault/ -SslProtocol TLS12`

**Comment by orionneo** [comment]
Hi Pete,

I do get an authentication failed, when I try to run this Invoke-WebRequest:

```
PS C:\Users\xxxxxa> Invoke-WebRequest -Uri https://myaPVWAaddress/PasswordVault/ -SslProtocol TLS12
Invoke-WebRequest : Authentication failed, see inner exception.
At line:1 char:1
+ Invoke-WebRequest -Uri https://rb-pam.bosch.com/PasswordVault/ -SslPr ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: GET, Re...rShell/6.1.0
}:HttpRequestMessage) [Invoke-WebRequest], HttpRequestException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

**Comment by orionneo** [comment]
Hi Pete,

In addition, I did the following on PS 5.0 and PSCore, please find results below:

**PS 5.0**
Code:
```
$url = "https://MyPVWAaddress/PasswordVault/"
$user = "username"
$pass= "password"
$pair = "$($user):$($pass)"
$encodedCredentials = [System.Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes($Pair)) >> diogo.txt
$headers = @{ Authorization = "Basic $encodedCredentials" } >> diogo.txt
Invoke-WebRequest -Uri $url -Method Get -Headers $headers -OutFile Config.html >> diogo.txt

```
Results:
```
emE3OGNhOiFEaWplMDAxOCoq

Name                           Value                                                                          
----                           -----                                                                          
Authorization                  Basic   
```                                                                       

**PS 6.0 (with -SslProtocol TLS12)**

Code:
```
$url = "https://MyPVWAaddress/PasswordVault/"
$user = "username"
$pass= "password"
$pair = "$($user):$($pass)"
$encodedCredentials = [System.Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes($Pair)) >> diogo.txt
$headers = @{ Authorization = "Basic $encodedCredentials" } >> diogo.txt
Invoke-WebRequest -Uri $url -Method Get -Headers $headers -OutFile Config.html -SslProtocol TLS12 >> diogo.txt

```

It tells me the authentication failed, but I also got this as output:

```
Invoke-WebRequest : Authentication failed, see inner exception.
emE3OGNhOiFEaWplMDAxOCoq

Name                           Value
----                           -----
Authorization                  Basic 
```

**Comment by orionneo** [comment]
@NathanielMaier , thanks for the hint! pw changed, as well as my last commands.

**Comment by pspete** [comment]
@orionneo - it appears the difference in behaviour you see in the module between PS5 & PSCore is that TLS1.2 is not currently enforced when using PSCore, which I will remedy in an upcoming release.
That being said - this update would mean you will likely see a TLS related error when using either PS5 or PSCore.

**Comment by orionneo** [comment]
Hi Pete,

Just to update you, I have re-installed the module after the last update, and it looks like it is working properly now.... any explanations?

**Comment by pspete** [comment]
Good to know your issue is fixed.
Between v2.2.2 & v2.3.0 of the module, the only thing which changed is that the `-SSLProtocol` parameter is specified on the `Invoke-WebRequest` call with a value of `TLS12` when the using PSCore. The rest of the updates all relate to new API features present in Cyberark 10.5. 
There were no changes to the logic of the authentication or web service call which would be used for PowerShell 5 at all.

**Comment by pspete** [comment]
@orionneo - I believe this is resolved for you now, so closing the issue.
